crates/wasmtime/src/runtime/ref/gc_ref.rs

```diff
@@ -1,52 +1,318 @@
+/// );
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
 #[repr(transparent)]
 pub struct ExternRef {
-    inner: VMExternRef,
+    inner: GcRootIndex,
 }
 
+unsafe impl GcRefImpl for ExternRef {}
+
 impl ExternRef {
     /// Creates a new instance of `ExternRef` wrapping the given value.
-    pub fn new<T>(value: T) -> ExternRef
+    ///
+    /// The resulting value is automatically unrooted when the given `context`'s
+    /// scope is exited. See [`Rooted<T>`][crate::Rooted]'s documentation for
+    /// more details.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn _foo() -> Result<()> {
+    /// let mut store = Store::<()>::default();
+    ///
+    /// // Create an `externref` wrapping a `str`.
+    /// let externref = ExternRef::new(&mut store, "hello!");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new<T>(mut context: impl AsContextMut, value: T) -> Rooted<ExternRef>
     where
         T: 'static + Any + Send + Sync,
     {
-        let inner = VMExternRef::new(value);
-        ExternRef { inner }
+        // Safety: We proviode `VMExternRef`'s invariants via the way that
+        // `ExternRef` methods take `impl AsContext[Mut]` methods.
+        let inner = unsafe { VMExternRef::new(value) };
+        Self::from_vm_extern_ref(context.as_context_mut().0, inner)
     }
 
-    pub(crate) fn from_vm_extern_ref(inner: VMExternRef) -> Self {
-        ExternRef { inner }
+    /// Creates a new, manually-rooted instance of `ExternRef` wrapping the
+    /// given value.
+    ///
+    /// The resulting value must be manually unrooted, or else it will leak for
+    /// the entire duration of the store's lifetime. See
+    /// [`ManuallyRooted<T>`][crate::ManuallyRooted]'s documentation for more
+    /// details.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmtime::*;
+    /// # fn _foo() -> Result<()> {
+    /// let mut store = Store::<()>::default();
+    ///
+    /// let externref = ExternRef::new_manually_rooted(&mut store, "hello!");
+    ///
+    /// // Use `externref` a bunch...
+    ///
+    /// // Don't forget to unroot the `externref` when done using it.
+    /// externref.unroot(&mut store);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new_manually_rooted<T>(
+        mut store: impl AsContextMut,
+        value: T,
+    ) -> ManuallyRooted<ExternRef>
+    where
+        T: 'static + Any + Send + Sync,
+    {
+        // Safety: We proviode `VMExternRef`'s invariants via the way that
+        // `ExternRef` methods take `impl AsContext[Mut]` methods.
+        let inner = unsafe { VMExternRef::new(value) };
+        let inner = extern_ref_into_gc_ref(inner);
+
+        // Safety: `inner` is a GC reference pointing to an `externref` GC
+        // object.
+        unsafe { ManuallyRooted::new(store.as_context_mut().0, inner) }
+    }
+
+    pub(crate) fn from_vm_extern_ref(store: &mut StoreOpaque, inner: VMExternRef) -> Rooted<Self> {
```

Comment:
I don't think this can be a safe function because there's no guarantee that `inner` belongs to `store`, right?